### PR TITLE
Topfails

### PR DESCRIPTION
**Pull request by jjchen**
Work on  https://github.com/gwicke/testreduce/issues/20 
Fail/skip/error extraction, score calculation, insertion into test_by_score.
Update in-memory queue of topfails in addResult.
Hook up front-end to display this list.

### CODE REVIEW

File: CassandraBackend.js [hunk 1]
```diff
@@ -344,6 +462,39 @@ CassandraBackend.prototype.addResult = function(test, commit, result, cb) {
...
+
+    // Update topFails
+    var index = findWithAttr(this.topFailsArray, "test", test);
+    if (index != -1 && this.topFailsArray[index].score <= score) {
+        this.topFailsArray[index].score = score;
+        this.topFailsArray[index].commit = commit;
+       // console.log("updated score");
+    }
+
+    this.topFailsArray.sort(function(a, b) { return b.score - a.score;} );
```
Comment (gwicke) [inline CassandraBackend.js]: I guess this could also be skipped if the score did not change.

File: CassandraBackend.js [hunk 2]
```diff
@@ -106,37 +122,102 @@ function getTests(cb) {
...
+                }
+            }
+
+            this.commitFails++;
+            if (this.commitFails < this.commits.length) {
+                var redo = initTopFails.bind( this );
+                redo(cb);
+            } else { 
+              //console.log("finished!: " + JSON.stringify(this.topFailsArray, null,'\t'));
+              async.sortBy(this.topFailsArray, function(fail, callback) {
```
Comment (gwicke) [inline CassandraBackend.js]: What's the reason for using async rather than a normal array sort here?

File: CassandraBackend.js [hunk 3]
```diff
@@ -360,24 +521,117 @@ var statsScore = function(skipCount, failCount, errorCount) {
...
+        if (err) {
+            console.log("err: " + err);
+            cb(err);
+        } else if (!results || !results.rows) {
+            console.log( 'no seen commits, error in database' );
+            cb(null);
+        } else {
+            //console.log("hooray we have data!: " + JSON.stringify(results, null,'\t'));
+            var fails = {};
+            async.each(results.rows, function(item, callback) {
```
Comment (gwicke) [inline CassandraBackend.js]: I think you can use result.rows.forEach here. The per-row handler is purely sync.

File: CassandraBackend.js [hunk 4]
```diff
@@ -360,24 +521,117 @@ var statsScore = function(skipCount, failCount, errorCount) {
...
+        if (err) {
+            console.log("err: " + err);
+            cb(err);
+        } else if (!results || !results.rows) {
+            console.log( 'no seen commits, error in database' );
+            cb(null);
+        } else {
+            //console.log("hooray we have data!: " + JSON.stringify(results, null,'\t'));
+            var skips = {};
+            async.each(results.rows, function(item, callback) {
```
Comment (gwicke) [inline CassandraBackend.js]: The callback is purely sync here, so you can just use results.rows.forEach.

**Comment by gwicke** [comment]
There are a few easy merge conflicts after #35 was merged. Could you resolve those as well?

File: CassandraBackend.js (hunk 2 above)
Comment (choucalate) [inline CassandraBackend.js]: I was going through some trial and error and this just happened to work better.  I could switch it back if it makes it simpler

File: CassandraBackend.js (hunk 1 above)
Comment (jjchen) [inline CassandraBackend.js]: Done.

File: CassandraBackend.js (hunk 3 above)
Comment (jjchen) [inline CassandraBackend.js]: Done.

File: CassandraBackend.js (hunk 4 above)
Comment (jjchen) [inline CassandraBackend.js]: Done.

**Comment by jjchen** [comment]
Fixed the above comments and merge conflicts.

**Comment by gwicke** [comment]
I didn't review everything super-carefully, but overall it looks good to me. Thanks!

**Pull request merged**